scalanet/src/io/iohk/scalanet/peergroup/InMemoryPeerGroup.scala

```diff
@@ -0,0 +1,142 @@
+    }
+}
+
+class InMemoryPeerGroup[A, M](address: A)(implicit network: Network[A, M]) extends PeerGroup[A, M] {
+
+  private var status: PeerStatus = PeerStatus.NotInitialized
+  private val channelStream = ConcurrentSubject[InMemoryChannel[A, M]](MulticastStrategy.publish)
+  private val channelsMap: TrieMap[ChannelID, InMemoryChannel[A, M]] = TrieMap()
+  def receiveMessage(channelID: ChannelID, from: A, msg: M): Unit = {
+    channelsMap.get(channelID) match {
+      case None =>
+        val newChannel = new InMemoryChannel(channelID, processAddress, from)
+        channelsMap += (channelID -> newChannel)
+        newChannel.depositMessage(msg)
+        channelStream.onNext(newChannel)
+      case Some(ch) =>
+        ch.depositMessage(msg)
+    }
+  }
+
+  // Public interface
+  def processAddress: A = address
+  def initialize(): Result[Unit] =
+    if (status == PeerStatus.Listening) error("Peer already connected")
+    else {
+      status = PeerStatus.Listening
+      network.register(this)
+    }
+  def client(to: A): Result[InMemoryChannel[A, M]] = {
+    val channelID = newChannelID()
+    val newChannel = new InMemoryChannel[A, M](channelID, processAddress, to)
+    channelsMap += (channelID -> newChannel)
+    allGood(newChannel)
+  }
+  def server(): Observable[InMemoryChannel[A, M]] = {
+    status match {
+      case PeerStatus.NotInitialized => throw new Exception(s"Peer $processAddress is not initialized yet")
+      case PeerStatus.Listening => channelStream
+      case PeerStatus.ShutDowned => throw new Exception(s"Peer $processAddress was shut downed")
+    }
+  }
+  def shutdown(): Result[Unit] = {
+    status match {
+      case PeerStatus.NotInitialized =>
+        allGood(()) // Should we return an error here?
+      case PeerStatus.Listening =>
+        status = PeerStatus.ShutDowned
+        network.disconnect(this)
+        allGood(())
+      case PeerStatus.ShutDowned =>
+        error("The peer group was already down") // Should we return an error here?
+    }
+  }
+}
+
+object InMemoryPeerGroup {
+
+  implicit def inMemoryPeerGroupTestUtils(implicit n: Network[Int, String]): PeerUtils[Int, String, InMemoryPeerGroup] =
+    PeerUtils.instance(
+      _.status == PeerStatus.Listening,
+      () => new InMemoryPeerGroup[Int, String](Random.nextInt()),
+      _.shutdown(),
+      _.initialize()
+    )
+
+  type ChannelID = java.util.UUID
+  def newChannelID(): ChannelID = java.util.UUID.randomUUID()
+
+  sealed trait PeerStatus
+  object PeerStatus {
+    case object NotInitialized extends PeerStatus
+    case object Listening extends PeerStatus
+    case object ShutDowned extends PeerStatus
+  }
+
+  trait ChannelStatus
+  object ChannelStatus {
+    case object Open extends ChannelStatus
+    case object Close extends ChannelStatus
+  }
+
+  type Result[A] = Task[A]
+  def allGood[A](x: A): Result[A] = Task.now(x)
+  def error(msg: String): Result[Nothing] = Task.now(throw new Exception(msg))
```

Comment:
Is this the correct way to initialize an always-failing task? I'm not very familiar with monix but I would expect something like `Future.failed`

Reply:
it seems that `Task.raiseError` can do the trick. I will try it.